src/index/CompressedRelation.cpp

```diff
@@ -458,51 +458,64 @@ auto getRelevantIdFromTriple(
                                maxId)
       .value_or(triple.col2Id_);
 }
-
-// `blockLessThanBlock` (a dummy) and `std::less<Id>` are only needed to
-// fulfill a concept for the `ql::ranges` algorithms.
-struct BlockLessThanBlock {
-  template <typename T = void>
-  bool operator()(const CompressedBlockMetadata&,
-                  const CompressedBlockMetadata&) const;
-};
 }  // namespace
 
 // _____________________________________________________________________________
-std::vector<CompressedBlockMetadata> CompressedRelationReader::getBlocksForJoin(
+auto CompressedRelationReader::getBlocksForJoin(
     ql::span<const Id> joinColumn,
-    const ScanSpecAndBlocksAndBounds& metadataAndBlocks) {
-  // We need symmetric comparisons between Ids and blocks.
+    const ScanSpecAndBlocksAndBounds& metadataAndBlocks)
+    -> GetBlocksForJoinResult {
+  if (joinColumn.empty() || metadataAndBlocks.getBlockMetadataView().empty()) {
+    return {};
+  }
+
+  // `id < block` iff `id < block.firstTriple`
   auto idLessThanBlock = [&metadataAndBlocks](
                              Id id, const CompressedBlockMetadata& block) {
     return id < getRelevantIdFromTriple(block.firstTriple_, metadataAndBlocks);
   };
 
+  // `block < id` iff `block.lastTriple < id`
   auto blockLessThanId = [&metadataAndBlocks](
                              const CompressedBlockMetadata& block, Id id) {
     return getRelevantIdFromTriple(block.lastTriple_, metadataAndBlocks) < id;
   };
 
-  auto lessThan = ad_utility::OverloadCallOperator{
-      idLessThanBlock, blockLessThanId, BlockLessThanBlock{}, std::less<Id>{}};
-
-  // Find the matching blocks by performing binary search on the `joinColumn`.
-  // Note that it is tempting to reuse the `zipperJoinWithUndef` routine, but
-  // this doesn't work because the implicit equality defined by
-  // `!lessThan(a,b) && !lessThan(b, a)` is not transitive.
-  auto blockIsNeeded = [&joinColumn, &lessThan](const auto& block) {
-    return !ql::ranges::equal_range(joinColumn, block, lessThan).empty();
-  };
-
   std::vector<CompressedBlockMetadata> result;
-  ql::ranges::copy(metadataAndBlocks.getBlockMetadataView() |
-                       ql::views::filter(blockIsNeeded),
-                   std::back_inserter(result));
+  const auto& mdView = metadataAndBlocks.getBlockMetadataView();
 
-  // The following check is cheap as there are only few blocks.
-  AD_CORRECTNESS_CHECK(std::unique(result.begin(), result.end()) ==
-                       result.end());
-  return result;
+  auto [colIt, colEnd] = getBeginAndEnd(joinColumn);
+  auto [blockIt, blockEnd] = getBeginAndEnd(mdView);
+  GetBlocksForJoinResult res;
+
+  // Manually count the blocks that we pass in the `mdView`.
```

Comment:
Please clarify that `pass` includes skipped blocks as well as returned blocks. Thinking about it, this is conceptually `std::ranges::distance(blockItOrig, blockIt)`.